app/src/ingest_edd_web.py

```diff
@@ -32,18 +32,47 @@ def _ingest_edd_web(
         urls_processed.add(item["url"])
 
         content = item.get("main_content", item.get("main_primary"))
-        assert content, f"Item {name} has no main_content or main_primary"
+        if not content:
+            logger.warning("Item %s has no main_content or main_primary", name)
+            continue
 
-        document = Document(name=name, content=content, **doc_attribs)
+        document = Document(name=name, content=content, source=item["url"], **doc_attribs)
         db_session.add(document)
 
-        embedding_model = app_config.sentence_transformer
-        tokens = len(embedding_model.tokenizer.tokenize(content))
-        mpnet_embedding = embedding_model.encode(content, show_progress_bar=False)
-        chunk = Chunk(
-            document=document, content=content, tokens=tokens, mpnet_embedding=mpnet_embedding
+        chunks = _chunk_page(content)
+        for chunk in chunks:
+            chunk.document = document
+        db_session.add_all(chunks)
+        logger.info("Split %s into %d chunks", name, len(chunks))
+
+
+def _chunk_page(content: str) -> list[Chunk]:
+    embedding_model = app_config.sentence_transformer
+
+    # Split content by double newlines, then gather into the largest chunks
+    # that tokenize to less than the max_seq_length
+    content_split_by_double_newlines = content.split("\n\n")
+    subsections = [content_split_by_double_newlines[0]]
+    for subsection in content_split_by_double_newlines[1:]:
+        tokens_with_new_subsection = len(
+            embedding_model.tokenizer.tokenize(subsections[-1] + subsection)
+        )
+        if tokens_with_new_subsection < embedding_model.max_seq_length:
+            subsections[-1] += "\n\n" + subsection
+        else:
+            subsections.append(subsection)
+
+    # Parallelize embedding generation for performance
```

Comment:
Suggest adding a log message so we get a sense of the number of chunks per webpage.

Reply:
There's a log statement after this function is called, so the log statements look like
> 2024-10-10 19:13:08,728 - INFO - Processing Quick Statistics Data Definitions (https://edd.ca.gov/en/about_edd/quick_statistics_definitions/)
2024-10-10 19:13:13,322 - INFO - Split Quick Statistics Data Definitions into 6 chunks